presto-main/src/main/java/com/facebook/presto/operator/aggregation/multimap/GroupedMultimapAggregationState.java

```diff
@@ -0,0 +1,76 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.multimap;
+
+import com.facebook.presto.array.ObjectBigArray;
+import com.facebook.presto.operator.aggregation.AbstractGroupCollectionAggregationState;
+import com.facebook.presto.operator.aggregation.TypedSet;
+import com.facebook.presto.operator.aggregation.arrayagg.MultimapAggregationStateConsumer;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+
+import static com.facebook.presto.operator.aggregation.multimap.MultimapUtils.addDistinctKeys;
+import static com.facebook.presto.operator.aggregation.multimap.MultimapUtils.writeSerializedMap;
+import static com.facebook.presto.type.TypeUtils.expectedValueSize;
+import static java.util.Objects.requireNonNull;
+
+public final class GroupedMultimapAggregationState
+        extends AbstractGroupCollectionAggregationState<MultimapAggregationStateConsumer>
+        implements MultimapAggregationState
+{
+    private static final int MAX_BLOCK_SIZE = 1024 * 1024;
+    private static final int VALUE_CHANNEL = 0;
+    private static final int KEY_CHANNEL = 1;
+
+    private final Type keyType;
+    private final Type valueType;
+
+    public GroupedMultimapAggregationState(Type keyType, Type valueType)
+    {
+        super(PageBuilder.withMaxPageSize(MAX_BLOCK_SIZE, ImmutableList.of(valueType, keyType)));
+        this.keyType = keyType;
+        this.valueType = valueType;
+    }
+
+    @Override
+    public final void add(Block keyBlock, int keyPosition, Block valueBlock, int valuePosition)
+    {
+        prepareAdd();
+        appendAtChannel(VALUE_CHANNEL, valueBlock, valuePosition);
+        appendAtChannel(KEY_CHANNEL, keyBlock, keyPosition);
+    }
+
+    @Override
+    protected final void accept(MultimapAggregationStateConsumer consumer, PageBuilder pageBuilder, int currentPosition)
+    {
+        consumer.accept(pageBuilder.getBlockBuilder(KEY_CHANNEL), pageBuilder.getBlockBuilder(VALUE_CHANNEL), currentPosition);
+    }
+
+    @Override
+    public final void toMultimapNativeEncoding(BlockBuilder blockBuilder)
+    {
+        requireNonNull(blockBuilder, "blockBuilder is null");
+
+        ObjectBigArray<BlockBuilder> valueArrayBlockBuilders = new ObjectBigArray<>();
+        int expectedSize = (int) Math.min(getPositionCount(), Integer.MAX_VALUE);
+        BlockBuilder distinctKeyBlockBuilder = keyType.createBlockBuilder(null, expectedSize, expectedValueSize(keyType, 100));
```

Comment:
nit: remove `expectedSize` local variable, and use `getEntryCount())`

Reply:
The result is this, which seems less readable:

```java
        BlockBuilder distinctKeyBlockBuilder = keyType.createBlockBuilder(null, (int) Math.min(getEntryCount(), Integer.MAX_VALUE), expectedValueSize(keyType, 100));
        TypedSet keySet = new TypedSet(keyType, (int) Math.min(getEntryCount(), Integer.MAX_VALUE), MultimapAggregationFunction.NAME);
```